hal/src/thumbv7em/clock/v2/rtcosc.rs

```diff
@@ -0,0 +1,207 @@
+impl Registers {
+    const OSC32KCTRL: *const RegisterBlock = crate::pac::OSC32KCTRL::ptr();
+
+    #[inline]
+    fn rtcctrl(&self) -> &RTCCTRL {
+        unsafe { &(*Self::OSC32KCTRL).rtcctrl }
+    }
+
+    #[inline]
+    fn set_source_freq(&mut self, source: DynRtcSourceId, freq: DynRtcFreq) {
+        use DynRtcFreq::*;
+        use DynRtcSourceId::*;
+        use RTCSEL_A::*;
+
+        let variant = match (source, freq) {
+            (OscUlp32k, Freq1k) => ULP1K,
+            (OscUlp32k, Freq32k) => ULP32K,
+            (Xosc32k, Freq1k) => XOSC32K,
+            (Xosc32k, Freq32k) => XOSC1K,
+        };
+        self.rtcctrl().write(|w| w.rtcsel().variant(variant));
+    }
+
+    fn reset_source_freq(&mut self) {
+        self.rtcctrl().reset();
+    }
+}
+
+//==============================================================================
+// RtcOscToken
+//==============================================================================
+
+pub struct RtcOscToken {
+    regs: Registers,
+}
+
+impl RtcOscToken {
+    pub(super) unsafe fn new() -> Self {
+        Self { regs: Registers }
+    }
+}
+
+//==============================================================================
+// RtcSourceId
+//==============================================================================
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum DynRtcSourceId {
+    OscUlp32k,
+    Xosc32k,
+}
+
+pub trait RtcSourceId {
+    const DYN: DynRtcSourceId;
+}
+
+impl RtcSourceId for OscUlp32kId {
+    const DYN: DynRtcSourceId = DynRtcSourceId::OscUlp32k;
+}
+
+impl RtcSourceId for Xosc32kId {
+    const DYN: DynRtcSourceId = DynRtcSourceId::Xosc32k;
+}
+
+//==============================================================================
+// RtcFreq
+//==============================================================================
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum DynRtcFreq {
+    Freq1k,
+    Freq32k,
+}
+
+pub trait RtcFreq: Sealed {
+    const DYN: DynRtcFreq;
+}
+
+pub enum Freq32k {}
+impl Sealed for Freq32k {}
+impl RtcFreq for Freq32k {
+    const DYN: DynRtcFreq = DynRtcFreq::Freq32k;
+}
+
+pub enum Freq1k {}
+impl Sealed for Freq1k {}
+impl RtcFreq for Freq1k {
+    const DYN: DynRtcFreq = DynRtcFreq::Freq1k;
+}
+
+//==============================================================================
+// RtcDriver
+//==============================================================================
+
+mod private {
+    use super::super::Driver;
+    use super::{DynRtcFreq, RtcSourceId};
+
+    pub trait RtcDriver: Driver
+    where
+        Self::Source: RtcSourceId,
+    {
+        fn enable_rtc_osc(&mut self, freq: DynRtcFreq);
+        fn disable_rtc_osc(&mut self, freq: DynRtcFreq);
+    }
+}
+
+use private::RtcDriver;
+
+impl<N> RtcDriver for Enabled<OscUlp32k, N>
+where
+    N: Counter,
+{
+    fn enable_rtc_osc(&mut self, freq: DynRtcFreq) {
+        use DynRtcFreq::*;
+        match freq {
+            Freq1k => self.0.token.activate_1k(true),
```

Comment:
As we spoke, 32k switch is also relevant for Gclks so this is not going to work (comment FTR)